tools/PI/DevHome.PI/Helpers/ExternalToolsHelper.cs

```diff
@@ -36,65 +68,161 @@ private ExternalToolsHelper()
+            // If the file has become corrupted, we should rename it (using DateTime.Now),
+            // create a new one, and report to the user. If it's locked, we just report to the user.
+            _log.Error(ex, "WriteToolsJsonFile unable to write to file");
         }
-
-        return tool;
     }
 
-    public void RemoveExternalTool(ExternalTool tool)
+    private void AllExternalTools_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
     {
-        if (externalTools.Remove(tool))
+        // Whenever the "all tools" collection changes, we need to synchronize the filtered collection.
+        switch (e.Action)
         {
-            // Write out to JSON file
-            var updatedJson = JsonSerializer.Serialize(externalTools, serializerOptions);
+            case NotifyCollectionChangedAction.Add:
+                if (e.NewItems is not null)
+                {
+                    foreach (ExternalTool newItem in e.NewItems)
+                    {
+                        if (newItem.IsPinned)
+                        {
+                            FilteredExternalTools.Add(newItem);
+                        }
 
-            try
-            {
-                File.WriteAllText(toolInfoFileName, updatedJson);
-            }
-            catch (Exception ex)
+                        newItem.PropertyChanged += ToolItem_PropertyChanged;
+                    }
+                }
+
+                break;
+
+            case NotifyCollectionChangedAction.Remove:
+                if (e.OldItems is not null)
+                {
+                    foreach (ExternalTool oldItem in e.OldItems)
+                    {
+                        oldItem.PropertyChanged -= ToolItem_PropertyChanged;
+                        FilteredExternalTools.Remove(oldItem);
+                    }
+                }
+
+                break;
+
+            case NotifyCollectionChangedAction.Replace:
+                if (e.OldItems is not null)
+                {
+                    foreach (ExternalTool oldItem in e.OldItems)
+                    {
+                        oldItem.PropertyChanged -= ToolItem_PropertyChanged;
+                        FilteredExternalTools.Remove(oldItem);
+                    }
+                }
+
+                if (e.NewItems is not null)
+                {
+                    foreach (ExternalTool newItem in e.NewItems)
+                    {
+                        if (newItem.IsPinned)
+                        {
+                            FilteredExternalTools.Add(newItem);
+                        }
+
+                        newItem.PropertyChanged += ToolItem_PropertyChanged;
+                    }
+                }
+
+                break;
+
+            case NotifyCollectionChangedAction.Reset:
+                SynchronizeAllFilteredItems();
+                break;
+        }
+    }
+
+    private void SynchronizeAllFilteredItems()
+    {
+        FilteredExternalTools.Clear();

```

Comment:
>FilteredExternalTools.Clear(); [](http://example.com/codeflow?start=8&length=30)

I wonder if it would make sense to build a collection first and then assign it to FilteredExternalTools at the end. Otherwise you'll be generated a bunch of changed events (no items, then 1 item, then 2 items, etc)... I bet the titlebar could potentially flicker based on system load?

Reply:
Ahh... yeah, I misread that.